ckanext/gdi_userportal/logic/action/fetcher/prop_fetcher.py

```diff
@@ -0,0 +1,38 @@
+from abc import ABC, abstractmethod
+
+from ckan.plugins import toolkit
+
+
+class PropFetcher(ABC):
+    def __init__(self, context, batch_size: int):
+        self._context = context
+        self._batch_size = batch_size
+
+    def get_prop_list(self) -> dict[str, int | list[str]]:
+        nb_datasets = self._get_dataset_count()
+        prop_values = []
+
+        for i in range((nb_datasets // self._batch_size) + 1):
+            batched_datasets = self._fetch_batched_datasets(
+                self._batch_size * i, self._batch_size
+            )
+            batched_prop_values = self._get_batched_prop_values(batched_datasets)
+            prop_values.extend(batched_prop_values)
+
+        return {"count": len(prop_values), "results": prop_values}
+
+    def _get_dataset_count(self) -> int:
+        return toolkit.get_action("package_search")(
+            self._context, {"include_private": False}
+        )["count"]
+
+    def _fetch_batched_datasets(self, start_row: int, nb_rows: int) -> list[dict]:
+        datasets = toolkit.get_action("package_search")(
```

Comment:
Now I understood what you meant about batch calls. 

I am worried about performance. `package_search` fetchs the entire dataset and then you cherry-pick one field, this is unfortunetally inefficient.

I advice you dig into how CKAN and fill the dataset search. There may be other APIs or internal services that could help you building that list directly. And `ckanext-scheming` is also another place to get your feet deep in mud.

The worst case, you can query directly into DB or SOLR, but ideally there should be a solution from CKAN or from `ckanext-scheming`.

 Have you check how CKAN list filters into the UI? I don't believe they do that, t

Reply:
By default, package_search fetches only **10 rows** (I should have made it explicit, or even passed the params). Consequently, retrieving the count of the dataset in the method `get_dataset_count` is quite cheap.

Actually, we can even pass `start=0&rows=0` to just retrieve the number of datasets (probably Solr is able to count efficiently with indexing and other optimizations).

I am not using `package_list` since it is not possible to filter out the private datasets.

I checked how CKAN proceeds with their filters but it is quite obscure. 
When accessing an organization, the URL becomes /organization/{org_name} and the page lists all the relevant filters.

Reply:
I am lack the knowledge of `package_search` response payload, as we are working with dictionaries only.

In this case, this code still some overhead but perhaps it doesn't seem to be a big concern for now. I advice we create a github issue to evaluate both design and performance concerns. Would you mind doing it?